blog/_src/posts/2018-08-16-if-you-give-a-mouse-a-macro.md

```diff
@@ -0,0 +1,445 @@
+just leave it at that and move on.
+
+Then, for each form `e` in a sequence, we can call `local-expand` with `def-ctx`
+and then check the expansion, `e-`, for `define/intermediate`. In those cases,
+we can use `int-def-ctx-bind-type-rename` to add it to the context. The
+procedure `add-bindings-to-ctx` performs this check on an expanded form `e-`:
+
+```racket
+(define-for-syntax (add-bindings-to-ctx e- def-ctx)
+  (syntax-parse e-
+        #:literals (erased define/intermediate)
+        [(erased (define/intermediate x:id x-:id τ e-))
+         (int-def-ctx-bind-type-rename #'x #'x- #'τ def-ctx)]
+        [_ (void)]))
+        
+```
+
+One extra detail that snuck in to the implementation is the presence of the
+`erased` tag. `erased` is a macro inserted by Turnstile that serves to mark
+language forms that have already been expanded, helping control what expands
+when. Since we used `define-typed-syntax` for our `define`, Turnstile wraps the
+output with `erased`.
+
+We now have the key ingredients to define a procedure, `walk/bind` that will
+serve as the primary vehicle to perform type checking of a sequence of forms,
+threading binding information through using a definition context. Processing
+sequences of defintions and expressions will iterate through them one at a time,
+and for each form `e`:
+
+- `local-expand` `e` using our internal definition context, resulting in an `e-`.
+- Retrieve the type of `e` from the metadata of `e-` using Turnstile's `typeof` helper.
+- Check if `e` defined a binding, in which case add it to the context.
+
+Aggregating the expanded syntax and type of each form as we go along, we get
+
+```racket
+(define-for-syntax (walk/bind e...)
+  (define def-ctx (syntax-local-make-definition-context))
+  (define unique (gensym 'walk/bind))
+  (define-values (rev-e-... rev-τ...)
+    (for/fold ([rev-e-... '()]
+               [rev-τ... '()])
+              ([e (in-syntax e...)])
+      (define e- (local-expand e (list unique) (list #'erased) def-ctx))
+      (define τ (typeof e-))
+      (add-bindings-to-ctx e- def-ctx)
+      (values (cons e- rev-e-...)
+              (cons τ rev-τ...))))
+  (values (reverse rev-e-...)
+          (reverse rev-τ...)))
+```
+
+The value `unique` and its use as an argument is dictated by the documentation
+of `local-expand`: "For a particular internal-definition context, generate a
+unique value and put it into a list for context-v." By using `#'erased` in the
+stop list for `local-expand`, we stop expansion at the same points that Turnstile
+does.
+
+Now we can use `walk/bind` to implement `begin`:
+
+```racket
+(define-typed-syntax (begin e ...+) ≫
+  #:do [(define-values (e-... τ...) (walk/bind #'(e ...)))]
+  #:with τ-final (last τ...)
+  --------------------
+  [⊢ (begin- #,@e-...) ⇒ τ-final])
+```
+
+and voilà!
+
+```racket
+(define (add2 [x : Int])
+  (define almost (+ x 1))
+  (+ almost 1))
+
+(add2 3)
+;;=> 5
+```
+
```

Comment:
A final runnable file of code at the end would likely make this much more useful to someone wanting to emulate this in the future.

Reply:
is putting it at the end preferable to where I have it at the beginning?